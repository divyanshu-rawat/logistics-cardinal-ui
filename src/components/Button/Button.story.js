import React, { Fragment } from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withInfo } from '@storybook/addon-info';
import { boolean, select } from '@storybook/addon-knobs';

import { Glyphicon } from 'react-bootstrap';

import withTests from '../../utils/withTests';

import { COMPONENTS } from '../../../.storybook/sections';

import Button from '.';

const generateButtonWithKnobs = (actionType, icon) => {
  const stretched = boolean('stretched', false);
  return (
    <div style={{ width: stretched ? '1000px' : 'auto' }}>
      <Button
        actionType={actionType}
        active={boolean('active', false)}
        disabled={boolean('disabled', false)}
        onClick={action((event) => console.log(event))}
        size={select('size', ['', 'large', 'small', 'xsmall'], '')}
        stretch={stretched}
      >
        <span>
          {icon && <Glyphicon glyph="star" />}
          My Awesome Button
        </span>
      </Button>
    </div>
  );
};

storiesOf(`${COMPONENTS}Button`, module)
  .addDecorator(withTests('Button/tests/index'))
  .add('Default', withInfo()(() => generateButtonWithKnobs()))
  .add('Primary', withInfo()(() => generateButtonWithKnobs('primary')))
  .add('Success', withInfo()(() => generateButtonWithKnobs('success')))
  .add('Info', withInfo()(() => generateButtonWithKnobs('info')))
  .add('Warning', withInfo()(() => generateButtonWithKnobs('warning')))
  .add('Danger', withInfo()(() => generateButtonWithKnobs('danger')))
  .add('Link', withInfo()(() => generateButtonWithKnobs('link')));

storiesOf(`${COMPONENTS}Button/With Icon`, module)
  .addDecorator(withTests('index'))
  .add('Default', withInfo()(() => generateButtonWithKnobs('default', true)))
  .add('Primary', withInfo()(() => generateButtonWithKnobs('primary', true)))
  .add('Success', withInfo()(() => generateButtonWithKnobs('success', true)))
  .add('Info', withInfo()(() => generateButtonWithKnobs('info', true)))
  .add('Warning', withInfo()(() => generateButtonWithKnobs('warning', true)))
  .add('Danger', withInfo()(() => generateButtonWithKnobs('danger', true)))
  .add('Link', withInfo()(() => generateButtonWithKnobs('link', true)));
