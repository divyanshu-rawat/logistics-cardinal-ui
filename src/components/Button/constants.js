export const SIZES = {
  XSMALL: 'xsmall',
  SMALL: 'small',
  DEFAULT: 'default',
  LARGE: 'large',
};

export const ACTION_TYPES = {
  DEFAULT: 'default',
  PRIMARY: 'primary',
  SUCCESS: 'success',
  INFO: 'info',
  WARNING: 'warning',
  DANGER: 'danger',
  LINK: 'link',
};
