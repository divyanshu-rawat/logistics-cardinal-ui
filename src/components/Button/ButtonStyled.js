import styled from 'styled-components';
import { darken, lighten } from 'polished';

import { SIZES } from './constants';

const buttonVariantSize = (
  fontSize,
  paddingY,
  paddingX,
  lineHeight,
  borderRadius,
) => `
  font-size: ${fontSize};
  padding: ${paddingY} ${paddingX};
  line-height: ${lineHeight};
  border-radius: ${borderRadius};
`;

const buttonVariantColor = (theme, key) => `
    border-color: ${theme[key]};
    background-color: ${theme[key]};

    &:focus,
    &:active {
      &:not(:disabled) {
        box-shadow: 0 0 0 0.2rem ${lighten(0.2, theme[key])};
      }
    }

    &:hover {
      &:not(:disabled) {
        border-color: ${darken(0.1, theme[key])};
        background-color: ${darken(0.1, theme[key])};
        border-color: ${darken(0.1, theme[key])};
      }
    }
  `;

const generateButtonSize = (
  {
    buttonFontSize,
    buttonFontSizeSmall,
    buttonFontSizeLarge,
    buttonPaddingY,
    buttonPaddingYXsmall,
    buttonPaddingYSmall,
    buttonPaddingYLarge,
    buttonPaddingX,
    buttonPaddingXXsmall,
    buttonPaddingXSmall,
    buttonPaddingXLarge,
    buttonLineHeight,
    buttonLineHeightSmall,
    buttonLineHeightLarge,
    buttonBorderRadius,
  },
  size,
) => {
  switch (size) {
    case SIZES.XSMALL:
      return buttonVariantSize(
        buttonFontSizeSmall,
        buttonPaddingYXsmall,
        buttonPaddingXXsmall,
        buttonLineHeightSmall,
        buttonBorderRadius,
      );
    case SIZES.SMALL:
      return buttonVariantSize(
        buttonFontSizeSmall,
        buttonPaddingYSmall,
        buttonPaddingXSmall,
        buttonLineHeightSmall,
        buttonBorderRadius,
      );
    case SIZES.LARGE:
      return buttonVariantSize(
        buttonFontSizeLarge,
        buttonPaddingYLarge,
        buttonPaddingXLarge,
        buttonLineHeightLarge,
        buttonBorderRadius,
      );
    default:
      return buttonVariantSize(
        buttonFontSize,
        buttonPaddingY,
        buttonPaddingX,
        buttonLineHeight,
        buttonBorderRadius,
      );
  }
};

export const ButtonStyled = styled('button')`
  display: inline-block;
  text-align: center;
  vertical-align: middle;
  user-select: none;
  background-color: transparent;
  font-weight: ${({ theme }) => theme.buttonFontWeight};
  color: ${({ theme }) => theme.buttonFontColor};
  border: ${({ theme }) => theme.buttonBorderWidth} solid transparent;
  width: ${({ block }) => (block ? '100%' : 'auto')};
  ${({ theme, size }) => generateButtonSize(theme, size)};

  &:focus,
  &.focus {
    outline: 0;
    box-shadow: ${({ theme }) => theme.buttonFocusBoxShadow};
  }

  &.disabled,
  &:disabled {
    box-shadow: none;
    cursor: not-allowed;
    opacity: ${({ theme }) => theme.buttonDisabledOpacity};
  }

  &:not(:disabled) {
    cursor: pointer;
  }

  &:focus,
  &:active {
    &:not(:disabled) {
      outline: 0;
    }
  }

  &:not(:disabled):active {
    box-shadow: ${({ theme }) => theme.buttonActiveBoxShadow};

    &:focus {
      box-shadow: ${({ theme }) => theme.buttonFocusBoxShadow};
    }
  }
`;

export const ButtonDefaultStyled = ButtonStyled.extend`
  color: ${({ theme }) => theme.primaryColor};
  border-color: ${({ theme }) => theme.primaryColor};

  &:hover:not(:disabled) {
    border-color: ${({ theme }) => darken(0.2, theme.primaryColor)};
  }

  &:focus:not(:disabled),
  &:active:not(:disabled) {
    box-shadow: 0 0 0 0.2rem ${({ theme }) => lighten(0.1, theme.primaryColor)};
  }
`;

export const ButtonPrimaryStyled = ButtonStyled.extend`
  ${({ theme }) => buttonVariantColor(theme, 'primaryColor')};
`;

export const ButtonSuccessStyled = ButtonStyled.extend`
  ${({ theme }) => buttonVariantColor(theme, 'successColor')};
`;

export const ButtonInfoStyled = ButtonStyled.extend`
  ${({ theme }) => buttonVariantColor(theme, 'infoColor')};
`;

export const ButtonWarningStyled = ButtonStyled.extend`
  ${({ theme }) => buttonVariantColor(theme, 'warningColor')};
`;

export const ButtonDangerStyled = ButtonStyled.extend`
  ${({ theme }) => buttonVariantColor(theme, 'dangerColor')};
`;
