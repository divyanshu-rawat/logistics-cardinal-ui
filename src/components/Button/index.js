import React from 'react';
import PropTypes from 'prop-types';
import {
  ButtonPrimaryStyled,
  ButtonSuccessStyled,
  ButtonDefaultStyled,
  ButtonDangerStyled,
  ButtonInfoStyled,
  ButtonWarningStyled,
} from './ButtonStyled';

import { ACTION_TYPES, SIZES } from './constants';

const resolveButtonType = (type) => {
  switch (type) {
    case ACTION_TYPES.PRIMARY:
      return ButtonPrimaryStyled;
    case ACTION_TYPES.SUCCESS:
      return ButtonSuccessStyled;
    case ACTION_TYPES.INFO:
      return ButtonInfoStyled;
    case ACTION_TYPES.WARNING:
      return ButtonWarningStyled;
    case ACTION_TYPES.DANGER:
      return ButtonDangerStyled;
    case ACTION_TYPES.LINK:
    default:
      return ButtonDefaultStyled;
  }
};

const Button = ({
  active,
  disabled,
  href,
  stretch,
  elementType,
  children,
  actionType,
  onClick,
  size,
}) => {
  const ButtonComponent = resolveButtonType(actionType);

  return (
    <ButtonComponent
      active={active}
      disabled={disabled}
      href={href}
      block={stretch}
      type={elementType}
      onClick={onClick}
      size={size}
    >
      {children}
    </ButtonComponent>
  );
};

Button.propTypes = {
  /**
   * The html inside of the button
   */
  children: PropTypes.any.isRequired,
  /**
   * It's going to generate the action color such as
   * .btn-primary, .btn-info, .btn-default and so on...
   */
  actionType: PropTypes.oneOf(Object.values(ACTION_TYPES)).isRequired,
  /**
   * Button type
   */
  elementType: PropTypes.oneOf(['button', 'reset', 'submit']).isRequired,
  /**
   * Flag to determine if the button is active or not
   */
  /**
   * Size of the button
   */
  size: PropTypes.oneOf(Object.values(SIZES)).isRequired,
  active: PropTypes.bool,
  /**
   * Flag to determine if the button is disabled or not
   */
  disabled: PropTypes.bool,
  /**
   * Anchor
   */
  href: PropTypes.string,
  /**
   * Makes button stretched
   */
  stretch: PropTypes.bool,
  /**
   * Action that will be triggered when the button is clicked
   */
  onClick: PropTypes.func,
};

Button.defaultProps = {
  active: false,
  disabled: false,
  href: '',
  stretch: false,
  actionType: ACTION_TYPES.DEFAULT,
  elementType: 'button',
  size: SIZES.DEFAULT,
};

export default Button;
