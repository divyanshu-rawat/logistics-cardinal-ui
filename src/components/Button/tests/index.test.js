import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Button from '..';

const onClickFn = jest.fn();

const renderedComponent = ({
  actionType = 'default',
  active = false,
  disabled = false,
  size = '',
  stretch = false,
}) =>
  shallow(
    <Button
      actionType={actionType}
      active={active}
      disabled={disabled}
      onClick={onClickFn}
      size={size}
      stretch={stretch}
    >
      Some Awesome Text
    </Button>,
  );

describe('<Button />', () => {
  describe('Default Button', () => {
    it('should render a default button', () => {
      const tree = toJson(renderedComponent({}));
      expect(tree).toMatchSnapshot();
    });

    it('should render a default button with different attributes from previous time', () => {
      const tree = toJson(
        renderedComponent({
          active: true,
          disabled: true,
          size: 'large',
          stretch: true,
        }),
      );
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Primary button', () => {
    it('should render a primary button', () => {
      const tree = toJson(renderedComponent({}));
      expect(tree).toMatchSnapshot();
    });

    it('should render a primary button with different attributes from previous time', () => {
      const tree = toJson(
        renderedComponent({
          active: true,
          disabled: true,
          size: 'large',
          stretch: true,
        }),
      );
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Success button', () => {
    it('should render a success button', () => {
      const tree = toJson(renderedComponent({}));
      expect(tree).toMatchSnapshot();
    });

    it('should render a success button with different attributes from previous time', () => {
      const tree = toJson(
        renderedComponent({
          active: true,
          disabled: true,
          size: 'large',
          stretch: true,
        }),
      );
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Info button', () => {
    it('should render a info button', () => {
      const tree = toJson(renderedComponent({}));
      expect(tree).toMatchSnapshot();
    });

    it('should render a info button with different attributes from previous time', () => {
      const tree = toJson(
        renderedComponent({
          active: true,
          disabled: true,
          size: 'large',
          stretch: true,
        }),
      );
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Warning button', () => {
    it('should render a warning button', () => {
      const tree = toJson(renderedComponent({}));
      expect(tree).toMatchSnapshot();
    });

    it('should render a warning button with different attributes from previous time', () => {
      const tree = toJson(
        renderedComponent({
          active: true,
          disabled: true,
          size: 'large',
          stretch: true,
        }),
      );
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Danger button', () => {
    it('should render a danger button', () => {
      const tree = toJson(renderedComponent({}));
      expect(tree).toMatchSnapshot();
    });

    it('should render a danger button with different attributes from previous time', () => {
      const tree = toJson(
        renderedComponent({
          active: true,
          disabled: true,
          size: 'large',
          stretch: true,
        }),
      );
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Link button', () => {
    it('should render a link button', () => {
      const tree = toJson(renderedComponent({}));
      expect(tree).toMatchSnapshot();
    });

    it('should render a link button with different attributes from previous time', () => {
      const tree = toJson(
        renderedComponent({
          active: true,
          disabled: true,
          size: 'large',
          stretch: true,
        }),
      );
      expect(tree).toMatchSnapshot();
    });
  });
});
