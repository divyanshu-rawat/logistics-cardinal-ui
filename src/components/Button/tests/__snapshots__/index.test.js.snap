// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Danger button should render a danger button 1`] = `
<ButtonStyled__ButtonDefaultStyled
  active={false}
  block={false}
  disabled={false}
  href=""
  onClick={[Function]}
  size="default"
  type="button"
>
  Some Awesome Text
</ButtonStyled__ButtonDefaultStyled>
`;

exports[`<Button /> Danger button should render a danger button with different attributes from previous time 1`] = `
<ButtonStyled__ButtonDefaultStyled
  active={true}
  block={true}
  disabled={true}
  href=""
  onClick={[Function]}
  size="large"
  type="button"
>
  Some Awesome Text
</ButtonStyled__ButtonDefaultStyled>
`;

exports[`<Button /> Default Button should render a default button 1`] = `
<ButtonStyled__ButtonDefaultStyled
  active={false}
  block={false}
  disabled={false}
  href=""
  onClick={[Function]}
  size="default"
  type="button"
>
  Some Awesome Text
</ButtonStyled__ButtonDefaultStyled>
`;

exports[`<Button /> Default Button should render a default button with different attributes from previous time 1`] = `
<ButtonStyled__ButtonDefaultStyled
  active={true}
  block={true}
  disabled={true}
  href=""
  onClick={[Function]}
  size="large"
  type="button"
>
  Some Awesome Text
</ButtonStyled__ButtonDefaultStyled>
`;

exports[`<Button /> Info button should render a info button 1`] = `
<ButtonStyled__ButtonDefaultStyled
  active={false}
  block={false}
  disabled={false}
  href=""
  onClick={[Function]}
  size="default"
  type="button"
>
  Some Awesome Text
</ButtonStyled__ButtonDefaultStyled>
`;

exports[`<Button /> Info button should render a info button with different attributes from previous time 1`] = `
<ButtonStyled__ButtonDefaultStyled
  active={true}
  block={true}
  disabled={true}
  href=""
  onClick={[Function]}
  size="large"
  type="button"
>
  Some Awesome Text
</ButtonStyled__ButtonDefaultStyled>
`;

exports[`<Button /> Link button should render a link button 1`] = `
<ButtonStyled__ButtonDefaultStyled
  active={false}
  block={false}
  disabled={false}
  href=""
  onClick={[Function]}
  size="default"
  type="button"
>
  Some Awesome Text
</ButtonStyled__ButtonDefaultStyled>
`;

exports[`<Button /> Link button should render a link button with different attributes from previous time 1`] = `
<ButtonStyled__ButtonDefaultStyled
  active={true}
  block={true}
  disabled={true}
  href=""
  onClick={[Function]}
  size="large"
  type="button"
>
  Some Awesome Text
</ButtonStyled__ButtonDefaultStyled>
`;

exports[`<Button /> Primary button should render a primary button 1`] = `
<ButtonStyled__ButtonDefaultStyled
  active={false}
  block={false}
  disabled={false}
  href=""
  onClick={[Function]}
  size="default"
  type="button"
>
  Some Awesome Text
</ButtonStyled__ButtonDefaultStyled>
`;

exports[`<Button /> Primary button should render a primary button with different attributes from previous time 1`] = `
<ButtonStyled__ButtonDefaultStyled
  active={true}
  block={true}
  disabled={true}
  href=""
  onClick={[Function]}
  size="large"
  type="button"
>
  Some Awesome Text
</ButtonStyled__ButtonDefaultStyled>
`;

exports[`<Button /> Success button should render a success button 1`] = `
<ButtonStyled__ButtonDefaultStyled
  active={false}
  block={false}
  disabled={false}
  href=""
  onClick={[Function]}
  size="default"
  type="button"
>
  Some Awesome Text
</ButtonStyled__ButtonDefaultStyled>
`;

exports[`<Button /> Success button should render a success button with different attributes from previous time 1`] = `
<ButtonStyled__ButtonDefaultStyled
  active={true}
  block={true}
  disabled={true}
  href=""
  onClick={[Function]}
  size="large"
  type="button"
>
  Some Awesome Text
</ButtonStyled__ButtonDefaultStyled>
`;

exports[`<Button /> Warning button should render a warning button 1`] = `
<ButtonStyled__ButtonDefaultStyled
  active={false}
  block={false}
  disabled={false}
  href=""
  onClick={[Function]}
  size="default"
  type="button"
>
  Some Awesome Text
</ButtonStyled__ButtonDefaultStyled>
`;

exports[`<Button /> Warning button should render a warning button with different attributes from previous time 1`] = `
<ButtonStyled__ButtonDefaultStyled
  active={true}
  block={true}
  disabled={true}
  href=""
  onClick={[Function]}
  size="large"
  type="button"
>
  Some Awesome Text
</ButtonStyled__ButtonDefaultStyled>
`;
